internal/app/machined/pkg/controllers/network/dns_server.go

```diff
@@ -0,0 +1,161 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/internal/pkg/ctxutil"
+	"github.com/siderolabs/talos/internal/pkg/dns"
+	"github.com/siderolabs/talos/pkg/machinery/resources/network"
+)
+
+// DNSServerController starts dns server on both udp and tcp ports based on finalized network configuration.
+type DNSServerController struct {
+	Addr string
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *DNSServerController) Name() string {
+	return "network.DNSServerController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *DNSServerController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.ResolverStatusType,
+			ID:        optional.Some(network.ResolverID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *DNSServerController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: network.DNSServerType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *DNSServerController) Run(originCtx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	handler := dns.NewHandler(logger)
+	defer handler.Stop()
+
+	cache := dns.NewCache(handler, logger)
+	addr := ctrl.addr()
+	ctx := originCtx
+
+	for _, opt := range []dns.ServerOptins{
+		{
+			Addr:    addr,
+			Net:     "udp",
+			Handler: cache,
+		},
+		{
+			Addr:          addr,
+			Net:           "tcp",
+			Handler:       cache,
+			ReadTimeout:   3 * time.Second,
+			WriteTimeout:  5 * time.Second,
+			IdleTimeout:   func() time.Duration { return 10 * time.Second },
+			MaxTCPQueries: -1,
+		},
+	} {
+		l := logger.With(zap.String("net", opt.Net))
+
+		runner := dns.NewRunner(dns.NewServer(opt), l)
+
+		err := ctrl.writeDNSStatus(ctx, r, opt.Net, handler)
+		if err != nil {
+			return err
+		}
+
+		// We attach here our goroutine to the context, so if goroutine exits for some reason,
+		// context will be canceled too.
+		ctx = ctxutil.MonitorFn(ctx, runner.Run)
+
+		defer ctrl.stopServer(r, handler, opt.Net, l)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctxutil.Cause(ctx)
+		case <-r.EventCh():
+		}
+
+		logger.Info("listing resolver status resources")
```

Comment:
this is crazy verbose and low information value

Reply:
Removed.